imports.go

```diff
@@ -100,69 +106,101 @@ func codeToPV(e *evaluator, filename string, code string) potentialValue {
+	return e.evaluate(pv)
 }
 
 // Concrete importers
 // -------------------------------------
 
-// FileImporter imports data from files.
+// FileImporter imports data from the filesystem.
 type FileImporter struct {
-	JPaths []string
+	JPaths  []string
+	fsCache map[string]*fsCacheEntry
+}
+
+type fsCacheEntry struct {
+	exists   bool
+	contents string
 }
 
-func tryPath(dir, importedPath string) (found bool, content []byte, foundHere string, err error) {
+func (importer *FileImporter) tryPath(dir, importedPath string) (found bool, content *string, foundHere string, err error) {
+	if importer.fsCache == nil {
+		importer.fsCache = make(map[string]*fsCacheEntry)
+	}
 	var absPath string
 	if path.IsAbs(importedPath) {
 		absPath = importedPath
 	} else {
 		absPath = path.Join(dir, importedPath)
 	}
-	content, err = ioutil.ReadFile(absPath)
-	if os.IsNotExist(err) {
-		return false, nil, "", nil
+	var entry *fsCacheEntry
+	if cacheEntry, isCached := importer.fsCache[absPath]; isCached {
+		entry = cacheEntry
+	} else {
+		contentBytes, err := ioutil.ReadFile(absPath)
+		if err != nil {
+			if os.IsNotExist(err) {
+				entry = &fsCacheEntry{
+					exists:   false,
+					contents: "",
+				}
+			} else {
+				return false, nil, "", err
+			}
+		} else {
+			entry = &fsCacheEntry{
+				exists:   true,
+				contents: string(contentBytes),
+			}
+		}
+		importer.fsCache[absPath] = entry
 	}
-	return true, content, absPath, err
+	return entry.exists, &entry.contents, absPath, nil
 }
 
-// Import imports a file.
-func (importer *FileImporter) Import(dir, importedPath string) (*ImportedData, error) {
-	found, content, foundHere, err := tryPath(dir, importedPath)
+// Import imports file from the filesystem.
+func (importer *FileImporter) Import(codePath, importedPath string) (contents *string, foundAt string, err error) {
+	dir, _ := path.Split(codePath)
+	found, content, foundHere, err := importer.tryPath(dir, importedPath)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 
 	for i := len(importer.JPaths) - 1; !found && i >= 0; i-- {
-		found, content, foundHere, err = tryPath(importer.JPaths[i], importedPath)
+		found, content, foundHere, err = importer.tryPath(importer.JPaths[i], importedPath)
 		if err != nil {
-			return nil, err
+			return nil, "", err
 		}
 	}
 
 	if !found {
-		return nil, fmt.Errorf("couldn't open import %#v: no match locally or in the Jsonnet library paths", importedPath)
+		return nil, "", fmt.Errorf("couldn't open import %#v: no match locally or in the Jsonnet library paths", importedPath)
 	}
-	return &ImportedData{Content: string(content), FoundHere: foundHere}, nil
+	return content, foundHere, nil
 }
 
 // MemoryImporter "imports" data from an in-memory map.
 type MemoryImporter struct {
-	Data map[string]string
+	Data map[string]*string
 }
 
-// Import imports a map entry.
-func (importer *MemoryImporter) Import(dir, importedPath string) (*ImportedData, error) {
+// Import fetches data from a map entry.
```

Comment:
The semantics of this are a bit weird now I think -- all paths are treated as absolute keys into the "Data" map.  That probably should be documented, although I understand we only use this for tests so it does not matter?